+++
title = "CellRanger Alignment Example"
date = "2022-07-25"
author = "Carly Martin"
categories = ["CellRanger", "Single-cell-genomics"]
tags = ["tak", "10x", "alignment", "cellranger"]
+++


Here, we align the NovaSeq outputs for 30k single nuclei profiles from DAP5 seeds using CellRanger. One 10x reaction can process 10k nuclei max, so the profiles are split across 3 reactions (DAP5A, DAP5B, DAP5C). CellRanger processes each reaction separately. 

## Organization and raw data naming:

We begin in the command line with the data copied from GTC storage to your solexa_gehring location, following this file structure: 
```
./cellranger/
    |-raw_data/
        |-220630_DAP5A_S1_L001_R1_001.fastq.gz
        |-220630_DAP5A_S1_L001_R2_001.fastq.gz
        |-220630_DAP5A_S1_L002_R1_001.fastq.gz
        |-220630_DAP5A_S1_L002_R2_001.fastq.gz
        |-220630_DAP5B_S2_L001_R1_001.fastq.gz
        |-220630_DAP5B_S2_L001_R2_001.fastq.gz
        |-220630_DAP5B_S2_L002_R1_001.fastq.gz
        |-220630_DAP5B_S2_L002_R2_001.fastq.gz
        |-220630_DAP5C_S3_L001_R1_001.fastq.gz
        |-220630_DAP5C_S3_L001_R2_001.fastq.gz
        |-220630_DAP5C_S3_L002_R1_001.fastq.gz
        |-220630_DAP5C_S3_L002_R2_001.fastq.gz
    |-DAP5A/  
    |-DAP5B/
    |-DAP5C/
    |-Arabidopsis_thaliana.TAIR10.43.gtf
    |-BSgenome.Athaliana.TAIR.TAIR9.fasta
```
Note that the raw data file names are not the names generated by illumina/GTC. CellRanger requires this naming convention:
```
[SAMPLE NAME]_S[X]_L00[LANE NUMBER]_[READ TYPE]_001.fastq.gz
```
Link with more info: https://kb.10xgenomics.com/hc/en-us/articles/115003802691-How-do-I-prepare-Sequence-Read-Archive-SRA-data-from-NCBI-for-Cell-Ranger-

Example of naming change, while copying to your raw_data folder in bash: 
```
#cd to your raw_data folder
cp /GTC/path/to/FASTQ/L224_1246_S1_L001_R1_001.fastq.gz ./220630_DAP5A_S1_L001_R1_001.fastq.gz 
```
## Preparing a custom reference for A. thaliana:

CellRanger requires a specialized directory structure for the reference genome and annotation. To construct a custom reference, prepare your .gtf file with cellranger mkgtf. Then, generate the reference file structure with cellranger mkref. 

In tak, call the v6 version of cellranger with cellranger6. The default version of cellranger (version 4) on tak cannot handle custom reference generation with plant genomes.

In bash, filter:
```
cellranger6 mkgtf Arabidopsis_thaliana.TAIR10.43.gtf Arabidopsis_thaliana.TAIR10.43.filtered.gtf
```
In bash, make the reference:
```
cellranger6 mkref --genome=TAIR9 --fasta=BSgenome.Athaliana.TAIR.TAIR9.fasta --genes=Arabidopsis_thaliana.TAIR10.43.filtered.gtf
```
Now, there is a new folder in your analysis directory containing the custom reference compatible with CellRanger:
```
./cellranger/
    |-raw_data/
        |-220630_DAP5A_S1_L001_R1_001.fastq.gz
        |-220630_DAP5A_S1_L001_R2_001.fastq.gz
        |-220630_DAP5A_S1_L002_R1_001.fastq.gz
        |-220630_DAP5A_S1_L002_R2_001.fastq.gz
        |-220630_DAP5B_S2_L001_R1_001.fastq.gz
        |-220630_DAP5B_S2_L001_R2_001.fastq.gz
        |-220630_DAP5B_S2_L002_R1_001.fastq.gz
        |-220630_DAP5B_S2_L002_R2_001.fastq.gz
        |-220630_DAP5C_S3_L001_R1_001.fastq.gz
        |-220630_DAP5C_S3_L001_R2_001.fastq.gz
        |-220630_DAP5C_S3_L002_R1_001.fastq.gz
        |-220630_DAP5C_S3_L002_R2_001.fastq.gz
    |-DAP5A/  
    |-DAP5B/
    |-DAP5C/
    |-Arabidopsis_thaliana.TAIR10.43.gtf
    |-BSgenome.Athaliana.TAIR.TAIR9.fasta
    |-TAIR9/
```
## Running the CellRanger alignment

Again, CellRanger processes one reaction at a time, so you must launch an alignment for each sample.

In bash, DAP5A:
```
cd ./raw_data/DAP5A

bsub -n 64 cellranger6 count --id=220630_whole_seed_DAP5A_colcol --transcriptome=TAIR9 --fastqs=../raw_data --sample=220630_DAP5A --expect-cells=10000 --localcores=8 --localmem=64
```
DAP5B:
```
cd ./raw_data/DAP5B

bsub -n 64 cellranger6 count --id=220630_whole_seed_DAP5B_colcol --transcriptome=/lab/solexa_gehring/carly/snRNA_seed_dev/cellranger/TAIR9 --fastqs=../raw_data --sample=220630_DAP5B --expect-cells=10000 --localcores=8 --localmem=64
```
DAP5C:
```
cd ./raw_data/DAP5C

bsub -n 64 cellranger6 count --id=220630_whole_seed_DAP5C_colcol --transcriptome=/lab/solexa_gehring/carly/snRNA_seed_dev/cellranger/TAIR9 --fastqs=../raw_data --sample=220630_DAP5C --expect-cells=10000 --localcores=8 --localmem=64

```
## Monitoring alignment progress

CellRanger provides a very useful link for visualizing the status of your alignment for each sample. To access the link for an alignment after you've submitted the cellranger count to the cluster, find the LSF job ID using the bjobs command, in bash:
```
bjobs
```
You'll see the job IDs as 7-digit numbers, for example: 6187243

Get detailed, instantaneous information about the job by using the bpeek command:
```
bpeek 6187243
```
When you run bpeek on a cellranger alignment, one of the first lines you will see in the output will contain a unique link, for example: http://c1b6:43737?auth=QNPwUNtdj7piPshR2dhGuzTRqy3PQKEA4w5Zp4QlRUQ

Copy and paste this into your browser to monitor the status of that run

## Locating QC files

When the run finishes, you will have the following files in your analysis directory (this is not a comprehensive structure, but contains the paths to the QC and the most important outs):
```
./cellranger/
    |-raw_data/
        |-220630_DAP5A_S1_L001_R1_001.fastq.gz
        |-220630_DAP5A_S1_L001_R2_001.fastq.gz
        |-220630_DAP5A_S1_L002_R1_001.fastq.gz
        |-220630_DAP5A_S1_L002_R2_001.fastq.gz
        |-220630_DAP5B_S2_L001_R1_001.fastq.gz
        |-220630_DAP5B_S2_L001_R2_001.fastq.gz
        |-220630_DAP5B_S2_L002_R1_001.fastq.gz
        |-220630_DAP5B_S2_L002_R2_001.fastq.gz
        |-220630_DAP5C_S3_L001_R1_001.fastq.gz
        |-220630_DAP5C_S3_L001_R2_001.fastq.gz
        |-220630_DAP5C_S3_L002_R1_001.fastq.gz
        |-220630_DAP5C_S3_L002_R2_001.fastq.gz

    |-DAP5A
        |-220630_whole_seed_DAP5A_colcol
            |-outs
                |-web_summary.html
                |-filtered_feature_bc_matrix
                    |-barcodes.tsv.gz
                    |-features.tsv.gz
                    |-matrix.mtx.gz    
    |-DAP5B
        |-220630_whole_seed_DAP5B_colcol
            |-outs
                |-web_summary.html
                |-filtered_feature_bc_matrix
                    |-barcodes.tsv.gz
                    |-features.tsv.gz
                    |-matrix.mtx.gz
    |-DAP5C
        |-220630_whole_seed_DAP5B_colcol
            |-outs
                |-web_summary.html
                |-filtered_feature_bc_matrix
                    |-barcodes.tsv.gz
                    |-features.tsv.gz
                    |-matrix.mtx.gz
```
Inspect the run QCs in the web_summary.html, and the filtered_feature_bc_matrix directory is the path you will use for loading the sample into programs like Seurat

ü•≥Ô∏è Happy clustering and plotting! ü•≥Ô∏è
